auth/src/source.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(dead_code)]
+
+use super::metadata;
+use crate::oauth2::{JwsClaims, JwsHeader};
+use crate::{AccessToken, Error, Result};
+use async_trait::async_trait;
+use chrono::{Duration, Utc};
+use rustls::sign::Signer;
+use rustls::sign::SigningKey;
+use rustls_pemfile::Item;
+use serde::{Deserialize, Serialize};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+const DEFAULT_HEADER: JwsHeader = JwsHeader {
+    alg: "RS256",
+    typ: "JWT",
+    kid: None,
+};
+const DEFAULT_OAUTH_GRANT: &str = "urn:ietf:params:oauth:grant-type:jwt-bearer";
+const DEFAULT_USER_GRANT: &str = "refresh_token";
+
+/// An producer of a [AccessToken].
+#[async_trait]
+pub trait Source {
+    async fn token(&self) -> Result<AccessToken>;
+}
+
+/// Options for building a [ServiceAccountKeySource].
+pub struct ServiceAccountKeySourceBuilder<'a> {
+    file_path: Option<PathBuf>,
+    contents: Option<&'a [u8]>,
+
+    scopes: Vec<String>,
+}
+
+impl ServiceAccountKeySourceBuilder<'_> {
+    pub fn scopes(mut self, scopes: Vec<String>) -> Self {
+        self.scopes.extend(scopes);
+        self
+    }
+    pub fn build(self) -> Result<ServiceAccountKeySource> {
+        let sa: ServiceAccountKeyFile = if let Some(contents) = self.contents {
+            serde_json::from_slice(contents)?
+        } else if let Some(file_path) = self.file_path {
+            serde_json::from_slice(&std::fs::read(file_path)?)?
+        } else {
+            return Err(Error::Other("".into()));
+        };
+        if self.scopes.is_empty() {
+            return Err(Error::Other("".into()));
+        }
+        Ok(ServiceAccountKeySource {
+            file: sa,
+            scopes: self.scopes,
+        })
+    }
+}
+
+/// A [Source] derived from a Service Account Key.
+pub struct ServiceAccountKeySource {
+    file: ServiceAccountKeyFile,
+    scopes: Vec<String>,
+}
+
+/// A representation of a Service Account File.
```

Comment:
If there is a public document describing the format of the key file, please link to it here.